waifu/modules/soda_vdm.py

```diff
@@ -0,0 +1,35 @@
+import typing as t
+
+from waifu.core.consts import PromptConstants
+from waifu.datasets.soda import SodaDataset
+from waifu.modules import BaseModule
+
+
+class SodaVDM(BaseModule):
+    '''Vanilla Dialogue Module based on the SODA dataset.'''
+
+    def generator(self) -> t.Generator[list[str], None, None]:
+        for episode in SodaDataset():
+            episode_messages = []
+            # Grab names to be replaced by <USER> and <BOT>. This is the first and second speaker respectively
+            # I'm assuming based on looking through dataset that it will always be user-bot-user-bot pattern.
+            user_name = episode.speakers[0]
+            bot_name = episode.speakers[1]
+            
+            # First, set the scenario.
+            # Make sure to replace any instance of first person in conversation with user token
+            replaced_narrative = episode.narrative.replace(user_name, PromptConstants.USER_TOKEN)
+            scenario = f"Scenario: {replaced_narrative}"
+            episode_messages.append(scenario)
+            # Next, the start token
+            episode_messages.append(PromptConstants.CHAT_START_TOKEN)
+            
+            # I am going to assume that the length of episode.speakers is the same as the length of episode.dialogue
+            # Looked pretty clean to me in the data. Fuck it, TODO: account for the possibility of that happening
+            for i, utterance in enumerate(episode.dialogue):
+                # For now, just leave bot's name unreplaced.
+                # If the index in the dialogue is even, it's the user talking (index 0), if not, it's the bot
+                name = PromptConstants.USER_PREFIX if i % 2 == 0 else bot_name
```

Comment:
Not always the case, I believe. The `speakers` array in the original dataset is a better way to pull this info (`name = episode.speakers[i]`, maybe?)

Reply:
That can work, yeah, though you would still need to replace the first speaker with the user token. A conditional can handle that.

Reply:
Looked through the changes you made. Couldn't find an example of non-sequential dialogue by quickly skimming through some rows though so I didn't test, but I think this should be fine for the vast majority of cases. Resolving.